### PR TITLE
Add SCM timestamp to JIRA comments

### PR DESCRIPTION
**Pull request by jan-zajic**
it's nice to have date and time of scm entries along with their author and id in comment messages.
It was proved by daily practice in our company. 
+ I refactor creation of one scm entry change description into method createScmChangeEntryDescription to be able to test it.

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/jira-plugin/83)

### CODE REVIEW

File: src/main/java/hudson/plugins/jira/Updater.java [hunk 1]
```diff
@@ -321,6 +281,79 @@ private String getScmComments(boolean wikiStyle,
         return comment.toString();
     }
 
+    protected String createScmChangeEntryDescription(Run<?, ?> run, Entry change, boolean wikiStyle,
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/Updater.java]: As you are refactoring already, could you try to split this big pile of code into two methods: wiki-style and non-wiki style? I think that would be a lot more readable and make introducing format changes easier.

Another way of doing that would be using some kind of templating to let the user define the format (still, the method split applies).

File: src/main/java/hudson/plugins/jira/Updater.java [hunk 2]
```diff
@@ -321,6 +281,79 @@ private String getScmComments(boolean wikiStyle,
...
+                }
+            }
+            description.append("\n");
+        }
+        return description.toString();
+    }
+
+    protected void appendChangeTimestampToDescription(StringBuilder description, JiraSite site, long timestamp) {
+        DateFormat df = null;
+        if (site.getDateTimePattern() != null) {
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/Updater.java]: I like this feature, but I think we need to make it optional, either preserve the former behavior when the format field is empty, or with a checkbox.

**Comment by jan-zajic** [comment]
ok, it's optional now (checkbox)

File: src/main/resources/hudson/plugins/jira/JiraSite/config.jelly [hunk 3]
```diff
@@ -35,6 +35,12 @@
   <f:entry title="${%Visible for Project Role}" field="roleVisibility">
     <f:textbox />
   </f:entry>
+  <f:entry title="${%Pattern for formatting date and time}" field="dateTimePattern">
+    <f:textbox />
+  </f:entry>
+  <f:entry title="${%Add timestamp to jira comments}" field="appendChangeTimestamp">
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraSite/config.jelly]: jira -> JIRA

File: src/main/resources/hudson/plugins/jira/JiraSite/help-appendChangeTimestamp.html [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<div>
+  If activated, scm change entries date and time will be recorded in jira.
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraSite/help-appendChangeTimestamp.html]: scm -> SCM

File: src/test/java/hudson/plugins/jira/UpdaterTest.java [hunk 5]
```diff
@@ -482,5 +506,60 @@ public void testGetChangesUsingReflectionForunknownJob() throws IOException {
...
+        when(r.getParent()).thenReturn(j);
+        JiraProjectProperty jiraProjectProperty = mock(JiraProjectProperty.class);
+        when(j.getProperty(JiraProjectProperty.class)).thenReturn(jiraProjectProperty);
+        when(jiraProjectProperty.getSite()).thenReturn(site);
+
+        ChangeLogSet.Entry entry = mock(ChangeLogSet.Entry.class);
+        when(entry.getTimestamp()).thenReturn(calendar.getTimeInMillis());
+        when(entry.getCommitId()).thenReturn("dsgsvds2re3dsv");
+        User mockAuthor = mock(User.class);
+        when(mockAuthor.getId()).thenReturn("jenkins-user");
```
Comment (warden) [inline src/test/java/hudson/plugins/jira/UpdaterTest.java]: move to constant and possibly use Mockito.anyString()

File: src/main/java/hudson/plugins/jira/Updater.java (hunk 2 above)
Comment (warden) [inline src/main/java/hudson/plugins/jira/Updater.java]: !completed

File: src/main/java/hudson/plugins/jira/Updater.java (hunk 1 above)
Comment (warden) [inline src/main/java/hudson/plugins/jira/Updater.java]: !completed

**Comment by warden** [comment]
Can you check the coverage for the refactored methods? It seems that some parts are not covered by the current tests.

File: src/main/java/hudson/plugins/jira/Updater.java [hunk 6]
```diff
@@ -321,6 +281,93 @@ private String getScmComments(boolean wikiStyle,
...
+        description.append("\n");
+    }
+
+    protected void appendChangeTimestampToDescription(StringBuilder description, JiraSite site, long timestamp) {
+        DateFormat df = null;
+        if (site.getDateTimePattern() != null) {
+            df = new SimpleDateFormat(site.getDateTimePattern());
+        } else {
+            // default format for current locale
+            df = DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.SHORT, Locale.getDefault());
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/Updater.java]: Add tests for case when no format is provided (e.g. when null or empty string)

**Comment by jan-zajic** [comment]
OK i increased test coverage and also try to fix some minor formatting issues, that i found during test writing. All existing tests are unchanged and passing.

File: src/main/resources/hudson/plugins/jira/JiraSite/config.jelly [hunk 7]
```diff
@@ -35,6 +35,12 @@
   <f:entry title="${%Visible for Project Role}" field="roleVisibility">
     <f:textbox />
   </f:entry>
+  <f:entry title="${%Pattern for formatting date and time}" field="dateTimePattern">
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraSite/config.jelly]: -> JIRA comments timestamp format

File: src/main/resources/hudson/plugins/jira/JiraSite/config.jelly [hunk 8]
```diff
@@ -35,6 +35,12 @@
...
+    <f:textbox />
+  </f:entry>
+  <f:entry title="${%Add timestamp to JIRA comments}" field="appendChangeTimestamp">
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraSite/config.jelly]: Move this before the pattern

File: src/test/java/hudson/plugins/jira/UpdaterTest.java [hunk 9]
```diff
@@ -310,10 +324,169 @@ public void testGetChangesUsingReflectionForWorkflowJob() throws IOException {
...
+        calendar.set(2016, 0, 1, 0, 0, 0);
+        JiraSite site = mock(JiraSite.class);
+        when(site.getDateTimePattern()).thenReturn("yyyy-MM-dd HH:mm:ss");
+        updater.appendChangeTimestampToDescription(description, site, calendar.getTimeInMillis());
+        System.out.println(description.toString());
+        Assert.assertThat(description.toString(), equalTo("2016-01-01 00:00:00"));
+    }
+
+    @Rule
+    public JenkinsRule rule = new JenkinsRule();
```
Comment (warden) [inline src/test/java/hudson/plugins/jira/UpdaterTest.java]: Move to beginning of class

File: src/main/resources/hudson/plugins/jira/JiraSite/help-dateTimePattern.html [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<div>
+See <a href="https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html">javadoc</a> for SimpleDateFormat for help.
+If not set, plugin gets the date/time formatter with the given formatting styles for the current locale.
```
Comment (warden) [inline src/main/resources/hudson/plugins/jira/JiraSite/help-dateTimePattern.html]: If not set, DateFormat.SHORT for the current locale will be used.

**Comment by warden** [comment]
Good work with the tests! Just small rewording/ordering changes and we'll be there.

**Comment by jglick** [comment]
@warden do you have any plans to cut a new release?

**Comment by warden** [comment]
Yeeeah, sorry, I was swamped a bit. I will do a release this week.

**Pull request merged**